core/app/models/spree/inventory_unit.rb

```diff
@@ -59,24 +69,81 @@ def order=(_)
 
     scope :shippable, -> { on_hand }
 
-    # state machine (see http://github.com/pluginaweek/state_machine/tree/master for details)
-    state_machine initial: :on_hand do
-      event :fill_backorder do
-        transition to: :on_hand, from: :backordered
-      end
-      after_transition on: :fill_backorder, do: :fulfill_order
+    def fill_backorder!
+      fill_backorder || (raise InvalidStateChange)
+    end
 
-      event :ship do
-        transition to: :shipped, if: :allow_ship?
-      end
+    def fill_backorder
+      return false unless can_fill_backorder?
+      change_state!(ON_HAND)
+      fulfill_order
+      true
+    end
 
-      event :return do
-        transition to: :returned, from: :shipped
-      end
+    def can_fill_backorder?
+      backordered?
+    end
 
-      event :cancel do
-        transition to: :canceled, from: CANCELABLE_STATES.map(&:to_sym)
-      end
+    def on_hand?
+      state == ON_HAND
+    end
+
+    def backordered?
+      state == BACKORDERED
+    end
+
+    def ship!
+      ship || (raise InvalidStateChange)
+    end
+
+    def ship
+      return false unless can_ship?
+      change_state!(SHIPPED)
+      true
+    end
+
+    def can_ship?
+      allow_ship?
+    end
+
+    def shipped?
+      state == SHIPPED
+    end
+
+    def return!
+      self.return || (raise InvalidStateChange)
+    end
+
+    def return
+      return false unless can_return?
+      change_state!(RETURNED)
+      true
+    end
+
+    def can_return?
+      shipped?
+    end
+
+    def returned?
+      state == RETURNED
+    end
+
+    def cancel!
+      cancel || (raise InvalidStateChange)
+    end
+
+    def cancel
+      return false unless can_cancel?
+      change_state!(CANCELED)
+      true
+    end
+
+    def can_cancel?
+      CANCELABLE_STATES.include?(state)
```

Comment:
Later we should change `CANCELABLE_STATES` to be a method as stores might want to change that.

Ideally this would even be configurable, so stores won't have to `class_eval` or `Module#prepend` this class